stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/EventHandlerSequential.kt

```diff
@@ -0,0 +1,797 @@
+
+    /**
+     * Checks if unread counts should be updated for particular channel.
+     * The unread counts should not be updated if channel in the DB
+     * does not contain [ChannelCapabilities.READ_EVENTS] capability.
+     *
+     * @param cid CID of the channel.
+     *
+     * @return True if unread counts should be updated
+     */
+    private suspend fun RepositoryFacade.hasReadEventsCapability(cid: String): Boolean {
+        return selectChannels(listOf(cid)).let { channels ->
+            val channel = channels.firstOrNull()
+            if (channel?.ownCapabilities?.contains(ChannelCapabilities.READ_EVENTS) == true) {
+                true
+            } else {
+                logger.d {
+                    "Skipping unread counts update for channel: $cid. ${ChannelCapabilities.READ_EVENTS} capability is missing."
+                }
+                false
+            }
+        }
+    }
+
+    private fun Message.enrichWithOwnReactions(batch: EventBatchUpdate, currentUserId: UserId, eventUser: User?) {
+        ownReactions = if (eventUser != null && currentUserId != eventUser.id) {
+            batch.getCurrentMessage(id)?.ownReactions ?: mutableListOf()
+        } else {
+            mergeReactions(
+                latestReactions.filter { it.userId == currentUserId },
+                batch.getCurrentMessage(id)?.ownReactions ?: mutableListOf()
+            ).toMutableList()
+        }
+    }
+
+    private fun GlobalMutableState.updateCurrentUser(me: User) {
+        _user.value = me
+        _mutedUsers.value = me.mutes
+        _channelMutes.value = me.channelMutes
+        _totalUnreadCount.value = me.totalUnreadCount
+        _channelUnreadCount.value = me.unreadChannels
+        _banned.value = me.banned
+    }
+
+    private infix fun UserId.mustBe(currentUserId: UserId?) {
+        if (this != currentUserId) {
+            throw InputMismatchException(
+                "received connect event for user with id $this while for user configured " +
+                    "has id $currentUserId. Looks like there's a problem in the user set"
+            )
+        }
+    }
+
+    companion object {
+        val EMPTY_DISPOSABLE = object : Disposable {
+            override val isDisposed: Boolean = true
+            override fun dispose() {}
+        }
+    }
+}
+
+private typealias UserId = String
+
+private class BatchEvent(
+    val sortedEvents: List<ChatEvent>,
+    val isFromHistorySync: Boolean,
+) {
+    val size: Int get() = sortedEvents.size
+    val isFromSocketConnection get() = !isFromHistorySync
+}
+
+private class SocketEventCollector(
+    private val scope: CoroutineScope,
+    private val fireEvent: suspend (BatchEvent) -> Unit
+) {
+    private val mutex = Mutex()
+    private val postponed = arrayListOf<ChatEvent>()
+    private val timeoutJob = AtomicReference<Job>()
+
+    suspend fun add(event: ChatEvent): Boolean {
+        if (event is UserStartWatchingEvent || event is UserStopWatchingEvent) {
```

Comment:
Should we crash or at least log an error in another case? 🤔

Reply:
We should not cause it collects just those 2 events for now. Other events are ignored.

Reply:
I would still add a log just to make sure it works as expected if someone decides to change the code